src/action.ml

```diff
@@ -736,37 +759,32 @@ let rec exec t ~ectx ~dir ~env_extra ~stdout_to ~stderr_to =
         (Marshal.to_string data [])
     in
     exec_echo stdout_to s
-  | Promote { mode; files } ->
-    let promote_mode = !Clflags.promote_mode in
-    if promote_mode = Ignore then
+  | Diff { optional; file1; file2 } ->
+    if (optional && not (Path.exists file1 && Path.exists file2)) ||
+       Io.read_file (Path.to_string file1) = Io.read_file (Path.to_string file2) then
```

Comment:
Isn't this a bit dangerous? If `optional` is false, then it might try to read files that don't exist or were deleted.

Reply:
That's not specific to diff.When executing an action, we always expect that all the dependencies are present.